### ISSUE TITLE
Tracker: Float16 support

### ISSUE THREAD

**Issue by maleadt** [JuliaGPU/CUDA.jl#391]
This issue is meant to be an overview / track progress on supporting Float16 in CUDA.jl

## Library APIs

We'll be able to do a lot already if we just dispatch to the correct library calls, e.g., `cublasHgemm` for half-precision, or `cublasGemmEx` with appropriate CUDA datatypes set. Note that this requires a matching math mode, https://docs.nvidia.com/cuda/cublas/index.html#cublassetmathmode, so https://github.com/JuliaGPU/CUDA.jl/issues/354 may be relevant.

Since we support many devices / toolkits, we'll need to check what the requirements for these calls are, and whether the APIs offer a fall-back (e.g., when tensor-cores are disabled or unavailable). If not, we may need to offer pure Julia fallbacks, convert to Float32 arrays, or require the caller (i.e. Knet, Flux) to check for device capabilities before using half-precision data. That may be the easiest solution.

Alternatively, we could look into the TF32 format which IIUC should make it possible to more easily use Float16 functions when available, or fall back to existing Float32 functionality (since the representation doesn't change). That way we also do not need kernel programming capabilities (see below) to implement non-library operations.

## Kernel programming

Beyond API calls, we need to be able to compile and execute half-precision code. This requires several features:

### Half-precision intrinsics

`libdevice` doesn't seem to provide half-precision intrinsics, but instead `cuda_fp16.hpp` contains inline assembly snippets:

```c++
__CUDA_FP16_DECL__ __half hsin(const __half a) {
    __half r = __hsin_internal(a);
    asm("{\n\t"
        "  .reg.b16 i,r,t;     \n\t"
        "  mov.b16 r, %0;      \n\t"
        "  mov.b16 i, %1;      \n\t"
        "  mov.b16 t, 0x8000;  \n\t"
        "  and.b16 t,r,t;      \n\t"
        __SPEC_CASE(i, r, 0X32B3, 0x0800)
        __SPEC_CASE(i, r, 0X5CB0, 0x1000)
        __SPEC_CASE(i, r, 0XB2B3, 0x8800)
        __SPEC_CASE(i, r, 0XDCB0, 0x9000)
        "  or.b16  r,r,t;      \n\t"
        "  mov.b16 %0, r;      \n"
        "}\n" : "+h"(__HALF_TO_US(r)) : "h"(__HALF_TO_CUS(a)));
    return r;
}
```

We can do inline assembly from Julia: https://github.com/JuliaGPU/CUDA.jl/blob/d9db2beeb6bb61f5d6b4c586b8de2e82dc43d78c/src/device/intrinsics/warp_vote.jl#L11-L21

### `half` LLVM representation

Julia currently represents half-precision floating-point as `i16` and implements essential operations itself. Ideally we'd switch to using `half` and let LLVM emit the appropriate GPU instructions, like https://github.com/JuliaLang/julia/pull/26381 does. Alternatively, we can use our own implementation, like https://github.com/vchuravy/IEEEFloat16.jl. @vchuravy, any status update here?

Note that we already can use LLVM's `half` with `llvmcall` after https://github.com/JuliaLang/julia/pull/33970; which makes it possible to, e.g., `llvmcall` the WMMA hardware intrinsics. We could use this to interface with other intrinsics, or even implement more essential operations, but it'd be better to update the Julia compiler and have it use `half` instead of reimplementing everything.

cc @denizyuret

**Comment by vchuravy** [comment]
Regarding intrinsics, one can also write readable Julia code instead of having to embed assembly: https://github.com/vchuravy/IEEEFloat16.jl/blob/547179b21aa56f19c161029488a1a95563f191a9/src/cuda.jl#L23-L29, but it is still a lot of work to implement and check everything. In the example above I haven't added the fixup checks, because Keno and I were confused about accuracy provided and the fixups made it worse at two points (IIRC).

Regarding switching all of Julia to `half` representation. Supporting `half` is not a requirement for any backend and while some like Intel have fallback lowering others like PPC didn't support it properly, although this might have been fixed by https://reviews.llvm.org/D68237. So we might be able to give that another try one LLVM11.

I am in favour of switching base Julia over to `half` and use hardware conversion ops instead of software ops through-out, but we need to keep it configurable :/ also it is annoying that CUDA support is dependent on the host supporting it as well.

Otherwise IEEEFloat16 is probably the way to go, but I can't devote time to it right now. I also found it funny that Nvidia is basically doing Float16 as a storage format + imprecise Float32 math.

**Comment by maleadt** [comment]
Ref https://github.com/JuliaGPU/CUDA.jl/pull/417, which adds Float16/BFloat16 and mixed-precision support to gemm

**Comment by maleadt** [comment]
And just to demonstrate why we need `half`/IEEEFloat16:

```
julia> add_one(x) = (@inbounds x[1] += one(eltype(x)); return)

julia> CUDA.code_ptx(add_one, Tuple{CuDeviceArray{Float32,0,AS.Global}})
//
// Generated by LLVM NVPTX Back-End
//

.version 6.3
.target sm_75
.address_size 64

        // .globl       julia_add_one_6987 // -- Begin function julia_add_one_6987
.weak .global .align 8 .u64 exception_flag;
                                        // @julia_add_one_6987
.visible .func julia_add_one_6987(
        .param .b64 julia_add_one_6987_param_0
)
{
        .reg .f32       %f<3>;
        .reg .b64       %rd<3>;

// %bb.0:                               // %top
        ld.param.u64    %rd1, [julia_add_one_6987_param_0];
        ld.u64  %rd2, [%rd1];
        ld.global.f32   %f1, [%rd2];
        add.f32         %f2, %f1, 0f3F800000;
        st.global.f32   [%rd2], %f2;
        ret;
                                        // -- End function
}
```

Nice and tidy. Let's try `Float16`:

```
julia> CUDA.code_ptx(add_one, Tuple{CuDeviceArray{Float16,0,AS.Global}})
//
// Generated by LLVM NVPTX Back-End
//

.version 6.3
.target sm_75
.address_size 64

        // .globl       julia_add_one_6995 // -- Begin function julia_add_one_6995
.extern .func  (.param .b32 func_retval0) julia_SUM__7000
(
        .param .b32 julia_SUM__7000_param_0,
        .param .b32 julia_SUM__7000_param_1
)
;
.weak .global .align 8 .u64 exception_flag;
                                        // @julia_add_one_6995
.visible .func julia_add_one_6995(
        .param .b64 julia_add_one_6995_param_0
)
{
        .reg .b32       %r<5>;
        .reg .b64       %rd<3>;

// %bb.0:                               // %top
        ld.param.u64    %rd1, [julia_add_one_6995_param_0];
        ld.u64  %rd2, [%rd1];
        ld.global.u16   %r1, [%rd2];
        mov.u32         %r2, 15360;
        { // callseq 25, 0
        .reg .b32 temp_param_reg;
        .param .b32 param0;
        st.param.b32    [param0+0], %r1;
        .param .b32 param1;
        st.param.b32    [param1+0], %r2;
        .param .b32 retval0;
        call.uni (retval0), 
        julia_SUM__7000, 
        (
        param0, 
        param1
        );
        ld.param.b32    %r3, [retval0+0];
        } // callseq 25
        st.global.u16   [%rd2], %r3;
        ret;
                                        // -- End function
}
```

welp. 

```
julia> using IEEEFloat16

julia> Base.one(::Type{IEEEFloat16.Float16}) = IEEEFloat16.Float16(convert(IEEEFloat16.Float16, 1f0))

julia> CUDA.code_ptx(add_one, Tuple{CuDeviceArray{IEEEFloat16.Float16,0,AS.Global}})
//
// Generated by LLVM NVPTX Back-End
//

.version 6.3
.target sm_75
.address_size 64

        // .globl       julia_add_one_7008 // -- Begin function julia_add_one_7008
.weak .global .align 8 .u64 exception_flag;
                                        // @julia_add_one_7008
.visible .func julia_add_one_7008(
        .param .b64 julia_add_one_7008_param_0
)
{
        .reg .b16       %h<4>;
        .reg .b64       %rd<3>;

// %bb.0:                               // %top
        ld.param.u64    %rd1, [julia_add_one_7008_param_0];
        ld.u64  %rd2, [%rd1];
        ld.global.b16   %h1, [%rd2];
        mov.b16         %h2, 0x3C00;
        add.f16         %h3, %h1, %h2;
        st.global.b16   [%rd2], %h3;
        ret;
                                        // -- End function
}
```

Much better. But the API coverage is insufficient. I wonder if we should try and use `IEEEFloat16` automatically via an Adapt rule.

**Comment by vchuravy** [comment]
> Much better. But the API coverage is insufficient. I wonder if we should try and use IEEEFloat16 automatically via an Adapt rule.

Yeah I think that should work, and API coverage is tedious but straightforward.

**Comment by maleadt** [comment]
I'm going to revisit `half` + that LLVM patch first.

**Comment by maleadt** [comment]
Looks like sm_80 supports more intrinsics, https://github.com/cupy/cupy/blob/620183256d25eb463081a8bac2a7a965d35db66b/cupy/_core/include/cupy/_cuda/cuda-12/cuda_fp16.hpp#L1462-L1475, so we should be specializing our overrides.

Those headers also define methods for more elementary stuff, like `+`. Does anybody know / We should check if we currently perform these natively already, or if we're doing these in Float32. I think these even are supported on older GPUs. I would hope that PTX handles this for us, or LLVM with ISEL based on the compute capability, but I haven't checked.
